packages/expo/src/useOAuth.ts

```diff
@@ -0,0 +1,84 @@
+import { useSignIn, useSignUp } from '@clerk/clerk-react';
+import type { OAuthStrategy } from '@clerk/types';
+import * as AuthSession from 'expo-auth-session';
+import * as WebBrowser from 'expo-web-browser';
+
+export type UseOAuthFlowParams =
+  | {
+      strategy: OAuthStrategy;
+      redirectUrl?: string;
+    }
+  | OAuthStrategy;
+
+export function useOAuth(params: UseOAuthFlowParams) {
+  let strategy: OAuthStrategy;
+  let redirectUrl: string;
+
+  if (typeof params === 'string') {
+    strategy = params;
+  } else if (typeof params === 'object') {
+    strategy = params.strategy;
+    redirectUrl = params.redirectUrl || '';
+  }
+
+  const { signIn, setSession, isLoaded: isSignInLoaded } = useSignIn();
+  const { signUp, isLoaded: isSignUpLoaded } = useSignUp();
+
+  async function startOAuthFlow() {
+    if (!isSignInLoaded || !isSignUpLoaded) {
+      return {};
+    }
+
+    // Create a redirect url for the current platform and environment.
+    //
+    // This redirect URL needs to be whitelisted for your Clerk production instance via
+    // https://clerk.dev/docs/reference/backend-api/tag/Redirect-URLs#operation/CreateRedirectURL
+    //
+    // For more information go to:
+    // https://docs.expo.dev/versions/latest/sdk/auth-session/#authsessionmakeredirecturi
+    const oauthRedirectUrl =
+      redirectUrl ||
+      AuthSession.makeRedirectUri({
+        path: '/oauth-native-callback',
+      });
+
+    await signIn.create({ strategy, redirectUrl: oauthRedirectUrl });
+
+    const { externalVerificationRedirectURL } = signIn.firstFactorVerification;
+
+    const result = await WebBrowser.openAuthSessionAsync(externalVerificationRedirectURL!.toString(), oauthRedirectUrl);
+
+    // @ts-expect-error
+    const { type, url } = result || {};
+
+    // TODO: Check all the possible AuthSession results
+    // https://docs.expo.dev/versions/latest/sdk/auth-session/#returns-7
+    if (type !== 'success') {
+      throw 'Something went wrong during the OAuth flow. Try again.';
+    }
+
+    const params = new URL(url).searchParams;
+
+    const rotatingTokenNonce = params.get('rotating_token_nonce') || '';
+    await signIn.reload({ rotatingTokenNonce });
+
+    const { status, firstFactorVerification } = signIn;
+
+    let createdSessionId = '';
+
+    if (status === 'complete') {
+      createdSessionId = signIn.createdSessionId!;
+    } else if (firstFactorVerification.status === 'transferable') {
+      await signUp.create({
```

Comment:
If we are creating the function we need to handle what happens if the user has signed up but requires additional fields, right now we aren't handling this in this flow. 

Are we expecting that to handle this by calling the function? Right now the code wouldn't complete.

I'd expect at bit more grace like

```
const redirectUrl = AuthSession.makeRedirectUri({
        path: "/oauth-native-callback",
      });

      await signIn.create({
        strategy: parsedStrat,
        redirectUrl,
      });

      const {
        firstFactorVerification: { externalVerificationRedirectURL },
      } = signIn;

      if (!externalVerificationRedirectURL)
        throw "Something went wrong during the OAuth flow. Try again.";

      const authResult = await AuthSession.startAsync({
        authUrl: externalVerificationRedirectURL.toString(),
        returnUrl: redirectUrl,
      });

      if (authResult.type !== "success") {
        throw "Something went wrong during the OAuth flow. Try again.";
      }

      // Get the rotatingTokenNonce from the redirect URL parameters
      const { rotating_token_nonce: rotatingTokenNonce } = authResult.params;

      await signIn.reload({ rotatingTokenNonce });

      const { createdSessionId } = signIn;

      if (createdSessionId) {
        // If we have a createdSessionId, then auth was successful
        await setSession(createdSessionId);
      } else {
        // If we have no createdSessionId, then this is a first time sign-in, so
        // we should process this as a signUp instead
        // Throw if we're not in the right state for creating a new user
        if (
          !signUp ||
          signIn.firstFactorVerification.status !== "transferable"
        ) {
          throw "Something went wrong during the Sign up OAuth flow. Please ensure that all sign up requirements are met.";
        }

        console.log(
          "Didn't have an account transferring, following through with new account sign up",
        );

        // Create user
        await signUp.create({ transfer: true });
        await signUp.reload({
          rotatingTokenNonce: authResult.params.rotating_token_nonce,
        });
        await setSession(signUp.createdSessionId);
      }
```

Reply:
Good point. I was thinking about all the cases and depending on the flow we have multiple outcomes for PSU sign ups, MFA in sign-ins, new security checks in the future during for both sign-in or sign-up, etc... 

The different outcomes are already modelled in signIn and signUp objects. Adding a conditional for all the combination will introduce maintenance overhead for the hook. 

So, the hook returns `{ createdSessionId, signIn, signUp, setSession}`. If createdSessionId exists then we are done.    Note that setSession is not called by the hook to give control to the developer to call it when they need to.

If createdSessionId doesn't exist,then the developer needs to use signIn or signUp, read their status and do the necessary handling.